CI/wininstaller/installer.iss

```diff
@@ -0,0 +1,733 @@
+  Result := UserDesktopPath;
+end;
+
+
+function InitializeSetup(): Boolean;
+begin
+  // Initialize the global variable during setup
+  GlobalUserName := GetCurrentSessionUserName();
+  GlobalUserDocsFolder := GetUserDocsFolder();
+  GlobalUserAppdataFolder := GetUserAppdataFolder();
+  
+  Result := True;
+end;
+
+
+function InitializeUninstall(): Boolean;
+begin
+  // Initialize the global variable during uninstall
+  GlobalUserName := GetCurrentSessionUserName();
+  GlobalUserDocsFolder := GetUserDocsFolder();
+  GlobalUserAppdataFolder := GetUserAppdataFolder();
+  
+  Result := True;
+end;
+
+
+procedure InitializeWizard();
+var
+  InstallPath, VCMIFolder, MapsFolder, DataFolder, Mp3Folder: String;
+begin
+  // Check if the application is already installed
+  IsUpgrade := RegQueryStringValue(HKCU, 'Software\VCMI', 'InstallPath', InstallPath);
+  if not IsUpgrade then
+  begin
+    // Create the install mode selection page only if it's not an upgrade
+    InstallModePage := CreateInputOptionPage(
+      wpWelcome,
+      ExpandConstant('{cm:SelectSetupInstallModeTitle}'),
+      ExpandConstant('{cm:SelectSetupInstallModeDesc}'),
+      ExpandConstant('{cm:SelectSetupInstallModeSubTitle}'),
+      True, False
+    );
+    
+    // Option 0
+    InstallModePage.Add(ExpandConstant(#13#10 + '  {cm:InstallForAllUsers}' + #13#10 + '   • {cm:InstallForAllUsers1}' + #13#10 + #13#10));
+    // Option 1
+    InstallModePage.Add(ExpandConstant(#13#10 + '  {cm:InstallForMeOnly}' + #13#10  +  '   • {cm:InstallForMeOnly1}' + #13#10 + '   • {cm:InstallForMeOnly2}' + #13#10));
+
+    if IsAdmin then
+    begin
+      // Default to "All Users"
+      InstallModePage.SelectedValueIndex := 0;
+    end
+    else
+    begin
+      // Default to "Me Only"
+      InstallModePage.SelectedValueIndex := 1;
+
+      // Disable the first option ("Install for All Users") for non-admins
+      InstallModePage.CheckListBox.ItemEnabled[0] := False;
+
+      // Force a redraw of the CheckListBox to fix appearance
+      InstallModePage.CheckListBox.Invalidate();
+    end;
+  end;
+ 
+  // Check for Heroes 3 installation paths
+  Heroes3Path := '';
+  Heroes3Path := RegistryQueryPath('SOFTWARE\GOG.com\Games\1207658787', 'path');
+  if Heroes3Path = '' then
+    Heroes3Path := RegistryQueryPath('SOFTWARE\WOW6432Node\GOG.com\Games\1207658787', 'path');
+  if Heroes3Path = '' then
+    Heroes3Path := RegistryQueryPath('SOFTWARE\New World Computing\Heroes of Might and Magic® III\1.0', 'AppPath');
+  if Heroes3Path = '' then
+    Heroes3Path := RegistryQueryPath('SOFTWARE\WOW6432Node\New World Computing\Heroes of Might and Magic® III\1.0', 'AppPath');
+  if Heroes3Path = '' then
+    Heroes3Path := RegistryQueryPath('SOFTWARE\New World Computing\Heroes of Might and Magic III\1.0', 'AppPath');
+  if Heroes3Path = '' then
+    Heroes3Path := RegistryQueryPath('SOFTWARE\WOW6432Node\New World Computing\Heroes of Might and Magic III\1.0', 'AppPath');
```

Comment:
in what form can inno consume constants from external files? C++ code contains the same paths in the launcher, I'd like to share them instead of copy-pasting.

Reply:
Inno Setup uses RegQueryStringValue in it's pascal scripting.

Reply:
And don't think I can link external function from non installed app. For example if we have this function in launcher I can use it from installer. Probably just system apis like I did for logged user name detection

Reply:
is it possible to "pre-extract" a file (say, to `%TEMP%`) and then reference it inside installer? E.g. a json or just something csv-like, then parse it to get the list of paths and keys.

Reply:
Question is why? Launcher have it hardcoded aswell.

Reply:
that's exactly the point: to get rid of hardcoding them in 2 places

Reply:
These strings will probably never changes. And even with another json, it means to distribute useless file with VCMI. Same as I wanted with GOG installers informations instead of hardcoding. And even with json, its likely impossible to deal it like this in Installer.

Reply:
ok, let's not touch it then